[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#74333 [FSDP] Override `named_parameters()` for clean names in `summon_full_params()`**

Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

Commit:
[FSDP] Clean param names in `summon_full_params()`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Clean param names in `summon_full_params()`"




[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1166,6 +1169,9 @@ def summon_full_params(
         """
+        # Monkey patch `named_parameters()`
+        torch_named_parameters = torch.nn.Module.named_parameters
+        self.named_parameters = self._fsdp_named_parameters  # type: ignore[assignment]
```
Adding the `type: ignore[assignment]` seems to be the best solution. See https://github.com/python/mypy/issues/2427.

Commit:
Update on "[FSDP] Clean param names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890 by monkey patching `torch.nn.Module.named_paramteters()`.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -10,10 +10,11 @@
-    List,
-    Optional,
     Generator,
+    Iterator,
```
Just sorted imports.

awgu commented:
To-Do: look into overriding `named_parameters()` instead of monkey patching, using the summon full params training state to indicate whether to use the special logic or simply call into the normal `nn.Module.named_parameters()`.

Commit:
Update on "[FSDP] Clean param names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890 by monkey patching `torch.nn.Module.named_paramteters()`.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1206,6 +1209,35 @@ def summon_full_params(
+        context manager.
+        """
+        # Determine which logic to use based on the context at call time
+        if not hasattr(self, "training_state") or \
```
`nn.Modules` contained in an `FSDP` instance may not have `training_state` as an attribute but still return `True` for `isinstance(module, FullyShardedDataParallel)`. If you guys think this `hasattr()` check is too hacky, I will see if I can find a more direct solution.

Commit:
Update on "[FSDP] Clean param names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890 by monkey patching `torch.nn.Module.named_paramteters()`.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1356,6 +1359,35 @@ def _free_full_params_and_use_local_shard(params_to_free):
 
+    def named_parameters(
+        self,
+        prefix: str = "",
```
Let's take in *args, **kwargs so we don't have to change if the API changes?

awgu replied:
Would it be correct to say that PyTorch Core cannot add positional arguments before `prefix` without breaking backward compatibility? In that case, could we only add `**kwargs`?

fegin replied:
Using `(*args, **kwargs)` is kind of the convention in PyThon to deal with this case.

awgu replied:
Oh, I misunderstood. I thought we were trying to _additionally_ pass in `*args, **kwargs` instead of replacing `prefix` and `recurse`.

Fixed this now.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1356,6 +1359,35 @@ def _free_full_params_and_use_local_shard(params_to_free):
+            ):
+                # Remove any instances of "_fsdp_wrapped_module._fpw_module.";
+                # there can be multiple in the case of nested FSDP modules
+                fsdp_prefix = FSDP_WRAPPED_MODULE + "." + FPW_MODULE + "."
```
Please create a FSDP_PREFIX global variable just like FSDP_WRAPPED_MODULE. This is actually a constant.

fegin reviewed (commented):
Some nits. Also one suggestion not related to the code, `named_parameter` should be mentioned in the PR title.

Commit:
Update on "[FSDP] Clean param names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

Commit:
Update on "[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

Commit:
Update on "[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

Commit:
Update on "[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1356,6 +1360,33 @@ def _free_full_params_and_use_local_shard(params_to_free):
+                # Remove any instances of the FSDP-specific prefix; there can
+                # be multiple in the case of nested FSDP modules
+                param_name = param_name.replace(FSDP_PREFIX, "")
+                yield (param_name, param)
```
Can we do the following to avoid the duplicated for loop?

```
in_summon = (training_state == summon_full_params)
for n, p in named_parameters():
    name = name.replace(...) if in_summon else name
    yield (name, p)
```

awgu replied:
Great point.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1356,6 +1360,33 @@ def _free_full_params_and_use_local_shard(params_to_free):
+        """
+        # Determine which logic to use based on the context at call time
+        if getattr(self, "training_state", None) != TrainingState_.SUMMON_FULL_PARAMS:
+            for param_name, param in torch.nn.Module.named_parameters(
```
Wonder if we can use `super().named_parameters` rather than torch.nn.Module.named_parameters call? If user writes `MyModule` that inherits from `nn.Module` with a custom named_parameters will this work?

awgu replied:
Great point.

rohan-varma reviewed (commented):
Looks good, just 2 minor q's for your consideration, will stamp after that

Commit:
Update on "[FSDP] Override `named_parameters()` for clean names in `summon_full_params()`"


Fixes https://github.com/pytorch/pytorch/issues/73890.

Differential Revision: [D34937201](https://our.internmc.facebook.com/intern/diff/D34937201)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34937201).

rohan-varma commented:
Lgtm

fegin reviewed (approved):
LGTM
